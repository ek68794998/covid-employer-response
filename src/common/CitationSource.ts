export class CitationSource {
	public date?: string;

	public link: string = "";

	public source: string = "";

	public title: string = "";
}
