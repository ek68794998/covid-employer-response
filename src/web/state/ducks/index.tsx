import { combineReducers } from "redux";

import { getLocalizedStrings } from "./localization/reducers";

export default combineReducers({
	strings: getLocalizedStrings,
});