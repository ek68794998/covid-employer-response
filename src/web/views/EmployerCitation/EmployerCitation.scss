$indicator-glow-size: 3px;
$indicator-size: 8px;

$indicator-negative: #f44336;
$indicator-neutral: #90a4ae;
$indicator-positive: #4caf50;

.indicator {
	border-radius: ($indicator-size / 2);
	display: inline-block;
	margin-right: 8px;
	height: $indicator-size;
	width: $indicator-size;

	&.indicator-negative {
		background: $indicator-negative;
		box-shadow:
			0 0 $indicator-glow-size $indicator-negative,
			0 0 $indicator-glow-size $indicator-negative;
	}

	&.indicator-neutral {
		background: $indicator-neutral;
		box-shadow:
			0 0 $indicator-glow-size $indicator-neutral,
			0 0 $indicator-glow-size $indicator-neutral;
	}

	&.indicator-positive {
		background: $indicator-positive;
		box-shadow:
			0 0 $indicator-glow-size $indicator-positive,
			0 0 $indicator-glow-size $indicator-positive;
	}
}

.citation-type {
	border-bottom: 1px dotted rgba(0, 0, 0, 0.5);
	cursor: help;
}

.citation-summary {
	margin-left: 4px;
}
