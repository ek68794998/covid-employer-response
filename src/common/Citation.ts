import { CitationSource } from "./CitationSource";
import { CitationType } from "./CitationType";

export class Citation {
	public positivity: number = 0;

	public sources?: CitationSource[];

	public summary: string = "";

	public type?: CitationType;
}
