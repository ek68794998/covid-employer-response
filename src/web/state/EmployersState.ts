import { EmployerRecord } from "../../common/EmployerRecord";

export interface EmployersState {
	items: EmployerRecord[];
}