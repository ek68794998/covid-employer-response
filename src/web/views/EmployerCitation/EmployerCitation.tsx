import React from "react";
import ReactMarkdown from "react-markdown";
import { RouteProps } from "react-router-dom";

import { Citation } from "../../../common/Citation";
import { CitationSource } from "../../../common/CitationSource";

import "./EmployerCitation.scss";

type IndicatorIconType = "add_circle" | "add_circle_outline" | "error_outline" | "remove_circle_outline" | "remove_circle";
type Neutrality = "Neutral" | "Positive" | "Negative";

interface Props extends RouteProps {
	citation: Citation;

	citationSourceBase: number;
}

const EmployerCitation: React.FC<Props> = (props: Props): React.ReactElement => {
	const { citation, citationSourceBase } = props;

	const getLinkComponent =
		(s: CitationSource, i: number): JSX.Element => {
			const date: Date | null = s.date ? new Date(s.date) : null;

			let title: string = s.title ? `${s.source}: ${s.title}` : s.source;

			if (date) {
				title = `${title} (${date.toLocaleDateString()} ${date.toLocaleTimeString()})`;
			}

			return <a key={i} href={s.link} rel="noopener noreferrer" target="_blank" title={title}>[{i + citationSourceBase}]</a>;
		};

	let indicatorClass: Neutrality = "Neutral";
	let indicatorIcon: IndicatorIconType = "error_outline";

	if (citation.positivity > 0) {
		indicatorClass = "Positive";
		indicatorIcon = citation.positivity > 1 ? "add_circle" : "add_circle_outline";
	} else if (citation.positivity < 0) {
		indicatorClass = "Negative";
		indicatorIcon = citation.positivity < -1 ? "remove_circle" : "remove_circle_outline";
	}

	const iconClass: string =
		`material-icons EmployerCitation__Indicator EmployerCitation__Indicator--${indicatorClass}`;

	return (
		<div className="EmployerCitation__Container">
			<i className={iconClass}>{indicatorIcon}</i>
			<span className="EmployerCitation__Summary">
				<ReactMarkdown source={citation.summary} />
				{citation.sources && <span className="EmployerCitation__References">{citation.sources.map(getLinkComponent)}</span>}
			</span>
		</div>
	);
};

export default EmployerCitation;
