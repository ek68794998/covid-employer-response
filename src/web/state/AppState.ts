import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}