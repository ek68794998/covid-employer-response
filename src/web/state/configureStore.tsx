import { AnyAction, applyMiddleware, compose, createStore, Store } from "redux";
import thunk from "redux-thunk";

import { Production } from "../../common/constants/EnvironmentConstants";

import { AppState } from "./AppState";
import rootReducer from "./ducks";

const configureStore = (preloadedState: Partial<AppState>): Store<AppState, AnyAction> => {
	const isProd: boolean = process.env.NODE_ENV === Production;
	const composeEnhancers: any =
		(!isProd && typeof window !== "undefined" && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

	const store: Store<AppState, AnyAction> = createStore(
		rootReducer,
		preloadedState,
		composeEnhancers(applyMiddleware(thunk)),
	);

	if (module.hot) {
		// Enable Webpack hot module replacement for reducers
		module.hot.accept("./ducks", () => {
			const nextRootReducer: any = require("./ducks").default;
			store.replaceReducer(nextRootReducer);
		});
	}

	return store;
};

export default configureStore;