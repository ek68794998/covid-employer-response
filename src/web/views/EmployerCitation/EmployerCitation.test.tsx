import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { Citation } from "../../../common/Citation";

import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import EmployerCitation from "./EmployerCitation";

describe("<EmployerCitation />", () => {
	test("renders without exploding", () => {
		const store: Store<AppState, AnyAction> = configureStore({});

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerCitation
							citation={new Citation()}
							citationSourceBase={0}
						/>
					</BrowserRouter>
				</Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});
});
