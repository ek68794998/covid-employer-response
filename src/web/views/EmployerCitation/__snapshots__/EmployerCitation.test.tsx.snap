// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployerCitation /> renders without exploding 1`] = `
<div
  className="EmployerCitation__Container"
>
  <i
    className="material-icons EmployerCitation__Indicator EmployerCitation__Indicator--Neutral"
  >
    error_outline
  </i>
  <span
    className="EmployerCitation__Summary"
  />
</div>
`;
